src/servers/src/http/jaeger.rs

```diff
@@ -690,50 +663,203 @@ fn error_response(err: Error) -> (HttpStatusCode, axum::Json<JaegerAPIResponse>)
     // maintain the mapping: trace_id -> (process_id -> service_name).
     let mut trace_id_to_processes: HashMap<String, HashMap<String, String>> = HashMap::new();
     // maintain the mapping: trace_id -> spans.
     let mut trace_id_to_spans: HashMap<String, Vec<Span>> = HashMap::new();
     // maintain the mapping: service.name -> resource.attributes.
     let mut service_to_resource_attributes: HashMap<String, Vec<KeyValue>> = HashMap::new();
 
+    let is_span_attributes_flatten = !records
+        .schema
+        .column_schemas
+        .iter()
+        .any(|c| c.name == SPAN_ATTRIBUTES_COLUMN);
+
     for row in records.rows.into_iter() {
         let mut span = Span::default();
-        let mut row_iter = row.into_iter();
+        let mut service_name = None;
+        let mut resource_tags = vec![];
+
+        for (idx, cell) in row.into_iter().enumerate() {
+            // safe to use index here
+            let column_name = &records.schema.column_schemas[idx].name;
+
+            match column_name.as_str() {
+                TRACE_ID_COLUMN => {
+                    if let JsonValue::String(trace_id) = cell {
+                        span.trace_id = trace_id.clone();
+                        trace_id_to_processes.entry(trace_id).or_default();
+                    }
+                }
+                TIMESTAMP_COLUMN => {
+                    span.start_time = cell.as_u64().context(InvalidJaegerQuerySnafu {
+                        reason: "Failed to convert timestamp to u64".to_string(),
+                    })? / 1000;
+                }
+                DURATION_NANO_COLUMN => {
+                    span.duration = cell.as_u64().context(InvalidJaegerQuerySnafu {
+                        reason: "Failed to convert duration to u64".to_string(),
+                    })? / 1000;
+                }
+                SERVICE_NAME_COLUMN => {
+                    if let JsonValue::String(name) = cell {
+                        service_name = Some(name);
+                    }
+                }
+                SPAN_NAME_COLUMN => {
+                    if let JsonValue::String(span_name) = cell {
+                        span.operation_name = span_name;
+                    }
+                }
+                SPAN_ID_COLUMN => {
+                    if let JsonValue::String(span_id) = cell {
+                        span.span_id = span_id;
+                    }
+                }
+                SPAN_ATTRIBUTES_COLUMN => {
+                    // for v0 data model, span_attributes are nested as a json
+                    // data structure
 
-        // Set trace id.
-        if let Some(JsonValue::String(trace_id)) = row_iter.next() {
-            span.trace_id = trace_id.clone();
-            trace_id_to_processes.entry(trace_id).or_default();
-        }
+                    if let JsonValue::Object(span_attrs) = cell {
+                        span.tags = object_to_tags(span_attrs);
```

Comment:
Let's use `span.tags.extend()` in case if the order is not ensured.

Reply:
There is only one `span_attributes` in the legacy data model

Reply:
https://github.com/GreptimeTeam/greptimedb/blob/c8db4b286d0bdc733e6a605416bcbcf7bfb6c919/src/servers/src/http/jaeger.rs#L826-L835

`scope_name`, `scope_version`, and some other tags are selected and pushed into the `span.tags` below, so it might have order issues if we use assignment here.

Reply:
I see. Let me fix this.